src/components/structures/RightPanel.js

```diff
@@ -17,17 +17,65 @@ limitations under the License.
 */
 
 import React from 'react';
+import PropTypes from 'prop-types';
 import { _t } from 'matrix-react-sdk/lib/languageHandler';
 import sdk from 'matrix-react-sdk';
-import Matrix from "matrix-js-sdk";
 import dis from 'matrix-react-sdk/lib/dispatcher';
 import MatrixClientPeg from 'matrix-react-sdk/lib/MatrixClientPeg';
 import Analytics from 'matrix-react-sdk/lib/Analytics';
 import rate_limited_func from 'matrix-react-sdk/lib/ratelimitedfunc';
-import Modal from 'matrix-react-sdk/lib/Modal';
 import AccessibleButton from 'matrix-react-sdk/lib/components/views/elements/AccessibleButton';
 import { showGroupInviteDialog } from 'matrix-react-sdk/lib/GroupInvite';
 
+class HeaderButton extends React.Component {
+    constructor() {
+        super();
+        this.onClick = this.onClick.bind(this);
+    }
+
+    onClick(ev) {
+        Analytics.trackEvent(...this.props.analytics);
+        dis.dispatch({
+            action: 'view_right_panel_phase',
+            phase: this.props.clickPhase,
+        });
+    }
+
+    render() {
+        const TintableSvg = sdk.getComponent("elements.TintableSvg");
+        const AccessibleButton = sdk.getComponent("elements.AccessibleButton");
+        const isHighlighted = this.props.phases.includes(this.props.currentPhase);
+
+        return <AccessibleButton
+            className="mx_RightPanel_headerButton"
+            key={this.props.key}
+            onClick={ this.onClick }
+        >
+            <div className="mx_RightPanel_headerButton_badge">
+                { this.props.badge ? this.props.badge : <span>&nbsp;</span>}
+            </div>
+            <TintableSvg src={this.props.iconSrc} width="25" height="25"/>
+            { isHighlighted ? <div className="mx_RightPanel_headerButton_highlight"></div> : <div/> }
+        </AccessibleButton>;
+    }
+}
+
+HeaderButton.propTypes = {
+    // If currentPhase is one of the specified phases, the button will be highlighted
+    phases: PropTypes.arrayOf(PropTypes.string).isRequired,
+    // The currentPhase of the RightPanel
+    currentPhase: PropTypes.string.isRequired,
```

Comment:
Is there likely to be any need for the button to know the actual phases it represents or what the current phase is? Could we just pass in whether it should be highlighted or not?

Reply:
>Could we just pass in whether it should be highlighted or not?

I was hoping to abstract this logic away such that, for example, we can't get two buttons highlighted at once. It's _probably_ overkill but if every single header button is going to embody this logic, it seems inefficient to dupe it for each one.

Reply:
Also, it needs to know for the dispatch so :man_shrugging: